metrics/density/footprint_data.sh

```diff
@@ -0,0 +1,354 @@
+# account for it
+function grab_dockerd_pss() {
+	if [[ $DUMP_HEADER ]]; then
+		echo -n ", dockerd_pss" >> $DATAFILE
+		return
+	fi
+
+	item=$(get_proc_pss $DOCKERD_PATH)
+	echo -n ", $item" >> $DATAFILE
+}
+
+# Get the PSS for the whole of userspace (all processes)
+#  This allows us to see if we had any impact on the rest of the system, for instance
+#  dockerd grows as we launch containers, so we should account for that in our total
+#  memory breakdown
+function grab_all_pss() {
+	if [[ $DUMP_HEADER ]]; then
+		echo -n ", all_pss" >> $DATAFILE
+		return
+	fi
+
+	item=$(sudo smem -t | tail -1 | awk '{print $5}')
+	item=$(($item * 1024))
+	echo -n ", $item" >> $DATAFILE
+}
+
+function grab_user_smem() {
+	if [[ $DUMP_HEADER ]]; then
+		echo -n ", user" >> $DATAFILE
+		return
+	fi
+
+	# userspace
+	item=$(sudo smem -w | head -5 | tail -1 | awk '{print $3}')
+	item=$(($item * 1024))
+	echo -n ", $item" >> $DATAFILE
+}
+
+function grab_slab() {
+	if [[ $DUMP_HEADER ]]; then
+		echo -n ", slab" >> $DATAFILE
+		return
+	fi
+
+	# Grabbing slab total from meminfo is easier than doing the math
+	# on slabinfo
+	item=$(cat /proc/meminfo | fgrep "Slab:" | awk '{print $2}')
+	item=$(($item * 1024))
+	echo -n ", $item" >> $DATAFILE
+}
+
+function get_memfree() {
+	mem_free=$(sudo smem -w | head -6 | tail -1 | awk '{print $4}')
+	mem_free=$(($mem_free * 1024))
+	echo $mem_free
+}
+
+function grab_system() {
+	if [[ $DUMP_HEADER ]]; then
+		echo -n ", avail" >> $DATAFILE
+		echo -n ", avail_incr" >> $DATAFILE
+		echo -n ", cached" >> $DATAFILE
+		echo -n ", free_smem" >> $DATAFILE
+		echo -n ", free_smem_incr" >> $DATAFILE
+		echo -n ", anon_pages" >> $DATAFILE
+		echo -n ", mapped_pages" >> $DATAFILE
+		echo -n ", cached" >> $DATAFILE
+
+		# Store up baseline measures
+		base_mem_avail=$(free -b | head -2 | tail -1 | awk '{print $7}')
+		base_mem_free=$(get_memfree)
+		return
+	fi
+
+	# avail memory, from 'free'
+	item=$(free -b | head -2 | tail -1 | awk '{print $7}')
+	echo -n ", $item" >> $DATAFILE
+	avail_incr=$(($base_mem_avail-$item))
+	echo -n ", $avail_incr" >> $DATAFILE
+
+	# cached memory, from 'free'
+	item=$(free -b | head -2 | tail -1 | awk '{print $6}')
+	echo -n ", $item" >> $DATAFILE
+
+	# free memory from smem
+	item=$(get_memfree)
+	echo -n ", $item" >> $DATAFILE
+	free_incr=$(($base_mem_free-$item))
+	echo -n ", $free_incr" >> $DATAFILE
+
+	# Anon pages
+	item=$(cat /proc/meminfo | fgrep "AnonPages:" | awk '{print $2}')
+	item=$(($item * 1024))
+	echo -n ", $item" >> $DATAFILE
+
+	# Mapped pages
+	item=$(cat /proc/meminfo | fgrep "Mapped:" | awk '{print $2}')
+	item=$(($item * 1024))
+	echo -n ", $item" >> $DATAFILE
+
+	# Cached
+	item=$(cat /proc/meminfo | grep "^Cached:" | awk '{print $2}')
+	item=$(($item * 1024))
+	echo -n ", $item" >> $DATAFILE
+}
+
+function grab_stats() {
+	# If configured, dump the caches so we get a more stable
+	# view of what our static footprint really is
+	if (($DUMP_CACHES)) ; then
```

Comment:
Did you mean `[[ $DUMP_CACHES ]]`?

Reply:
well, I explicitly put in the comments that it should be set to 0 or 1 :-) - but, I'll flip it to set or not...

Reply:
I was actually unfamiliar with the `(( ))` operator (embarrassing, since it's right above `[[ ]]` in the manpage). I just saw weird behavior with one test case, so I wasn't sure it was doing the right thing:
    btwarden@btwarden-ws:~$ DUMP_CACHES="0" if (($DUMP_CACHES)); then echo true; else echo false; fi
    bash: syntax error near unexpected token `('

But these test cases worked as expected, so your code is probably best:
    btwarden@btwarden-ws:~$ export DUMP_CACHES=0
    btwarden@btwarden-ws:~$ if (($DUMP_CACHES)); then echo true; else echo false; fi
    false
    btwarden@btwarden-ws:~$ export DUMP_CACHES=1
    btwarden@btwarden-ws:~$ if (($DUMP_CACHES)); then echo true; else echo false; fi
    true
    btwarden@btwarden-ws:~$ unset DUMP_CACHES
    btwarden@btwarden-ws:~$ if (($DUMP_CACHES)); then echo true; else echo false; fi
    false
    btwarden@btwarden-ws:~$ export DUMP_CACHES=foo
    btwarden@btwarden-ws:~$ if (($DUMP_CACHES)); then echo true; else echo false; fi
    false